### DISCUSSION TITLE
Is it possible to add another dev directory inside /app for react-rails?

### DISCUSSION THREAD

**Discussion by bubuluxWork** [reactjs/react-rails#1343] [Q&A] [answered]
I am migrating our codebase to Typescript and want new code to be written in a new directory: app/typescript. Note: I don't need any SSR, typical SPA behavior is sufficient. It would be nice to be able to configure bundle splitting (like with app/javascript/packs)...

I am very new to the entire RoR ecosystem and thus it's tough for me to pinpoint where I want to look for making changes. 

Would anyone be able to help me out here? Is what I want even possible?

In any case: Thx for all the resources you guys invest in making RoR development approachable for people with vastly different backgrounds (React etc.), really appreciated :)

Best, Sam

**Comment by ritikm31** [comment, marked as answer]
Yes, it’s possible to add another dev directory inside the /app folder in a React-Rails application, but it depends on your specific use case. Here’s a general overview of how you might approach it:

Creating the Directory: You can add a new directory under /app by simply creating an app/dev folder. This is useful for organizing files or components specific to your development environment, like development-only React components, utilities, or test helpers.

Configuring Webpack (if you're using it): If the dev directory contains JavaScript files or React components, you may need to update your Webpack configuration (if custom-configured) to ensure it compiles assets from this new directory.

Rails Autoload Paths: By default, Rails autoloads certain directories within /app like /app/models, /app/controllers, and /app/helpers. If you need Rails to load Ruby files from the dev directory automatically, you can configure it by adding this line to your config/application.rb:

ruby
Copy code
config.autoload_paths += %W(#{config.root}/app/dev)
Environment-Specific Code: If you only want the code in app/dev to be available in development, you can wrap it with Rails.env.development? checks.

React-Rails Integration: If you're using React components in this dev directory and you want Rails to recognize them, make sure the paths are accessible. The react-rails gem will load components based on what is available to the asset pipeline or Webpack configuration.

**Reply by bubuluxWork** [reply]
Sorry, that was a while back, just saw the reply. Thx for taking the time to explain this, I eventually stuck with the defaults. Thx anyhow

**Comment by aniruddhaadak80** [comment]
Hi there! Taking a look at your issue regarding `Is it possible to add another dev directory inside /app for react-rails?` in the context of `reactjs/react-rails`...

In the `react-rails` ecosystem, this generally traces back to the asset pipeline (Sprockets vs Webpack/esbuild/Vite) interacting with modern React.
1. **Bundling:** If you're encountering hydration mismatches or module loading issues, verify that your entry points in `app/javascript/components` properly export default/named components as expected by the `react_component` helper.
2. **SSR:** If server-side rendering is enabled, ensure no browser-specific APIs (`window`, `document`) are invoked during the initial render phase in your components. Move those inside `useEffect`.
3. **Caches:** A standard recommendation is running `rake tmp:clear` and clearing out `public/packs` or `public/assets` to be absolutely certain it's not a stale bundle.

Hope this gets your Rails/React setup unblocked and running smoothly!
